server/src/main/java/org/elasticsearch/common/document/DocumentField.java

```diff
@@ -47,14 +46,16 @@
 public class DocumentField implements Streamable, ToXContentFragment, Iterable<Object> {
 
     private String name;
+    private Boolean isMetadataField;
```

Comment:
can this be `final`?

Reply:
We need to think more how to organize `DocumentField` class.  Usually `fromXContent -> toXContent ->fromXContent` should produce an equal object (and this is what we test in `DocumentFieldTests`). The way how you organized a `DocumentField` class, it doesn't happen as `toXContent` is using `isMetadataField`, and `fromXConent` is not using it.

Reply:
if it requires a lot of changes to use `isMetadataField` in `toXContent` (we need to investigate this first), then we should document why we are not including it in `toXContent`, and also exclude  this field from `equals`, `hashCode` - also documenting why `isMetadataField`  is not participating in these functions

Reply:
Thanks for the review! 

Regarding consistency of ```fromXContent -> toXContent ->fromXContent ``` we had a relevant discussion in the issue description. There was a suggestion of using different json structures  depending when serializing from new versions, while still being able to deserialize json from old versions, as suggested in https://github.com/elastic/elasticsearch/issues/24422#issuecomment-455913816  . In such case it would have been possible to have enough information in the json to be able to make the conversions consistent without context. Yet there was a strong argument to avoid as mentioned here https://github.com/elastic/elasticsearch/issues/24422#issuecomment-456632130
as changing such json structure may result in issues when serializing it in one version and parsing in another.

In short, looks like it is not a lot of changes, but a bwc concern that prevents us from storing the `isMetadataField` in `toXContent`. It would be great if you share your thoughts on the matter.

Reply:
here we need to make a trade-off between a)extending serialized json schema for document field, additional coding to guarantee backward compatibility for serialized content and b)guessing the meaning of json fields depending on the context, which may potentially introduce some bugs if our assumptions are incorrect in some case, e.g. it is not completely obvious if it is a right approach here https://github.com/elastic/elasticsearch/pull/38373/files/314c58a64e5c4f2fb3fc9ec6e3366a0209597dd5#r253675646 

what is the process for deciding between `a` and `b` here?

Reply:
@Sandmannn Thanks for clarification. Ok, lets leave your code for json serialization for `DocumentField` at it is now, but just put comments why `isMetadataField` is a special field, why it is not participating in equals, hashCode, toString, and `toXContent`.
You still need to modify `readFrom` and `writeTo` methods to include this field.